### PR TITLE
refactor: strict typescript

### PR DESCRIPTION
**Pull request by TimBeyer**
**What this PR does / why we need it**:
This PR puts TS into strict mode.
I had to refactor a whole bunch of types, especially since we're using properties on classes as a way to pass type information along. I manage to get rid of that in most but not all cases.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by stefreak — COMMENTED** [review]
Only one small comment, otherwise looks awesome! ❤️

File: core/src/cli/cli.ts [hunk 1]
```diff
@@ -91,12 +91,12 @@ export class GardenCli {
       .sort()
       .filter((cmd) => cmd.getPath().length === 1)
 
-    let msg = `
-${cliStyles.heading("USAGE")}
-  garden ${cliStyles.commandPlaceholder()} ${cliStyles.optionsPlaceholder()}
+    let msg = dedent`
+      ${cliStyles.heading("USAGE")}
```
Comment (stefreak) [inline core/src/cli/cli.ts:95]: Oh, maybe this was not indented due to the ANSI code bug in dedent?

Comment (TimBeyer) [reply core/src/cli/cli.ts:95]: Oh that makes sense. Damn

**Review by stefreak — APPROVED** [review]
Wow, how did you do manage to push through and enable strict mode? Awesome work 🥇 ❤️ 
LGTM

**Pull request merged**